be/src/exec/pipeline/context_base.h

```diff
@@ -0,0 +1,63 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021 StarRocks Limited.
+
+#pragma once
+
+#include "common/status.h"
+#include "runtime/runtime_state.h"
+
+namespace starrocks::pipeline {
+
+// The complex ExecNode is usually decomposed to multiple operators,
+// which share some variables in a context, and have dependency relationships between them.
+// (For example, HashJoinNode is composed to HashJoinBuildOperator and HashJoinProbeOperator,
+// which share a HashJoiner context. And HashJoinProbeOperator depends on HashJoinBuildOperator.)
+// Assume OpA depends on OpB.
+// However, OpA can be finished earlier than OpB, if it's successor operator is finished early.
+// Therefore, there are two problems which needs to be considered:
+// 1. OpB should perceive that OpA has been finished early.
+// 2. Context object can be released only if both OpA and OpB are closed,
+//    because the close order of OpA and OpB are uncertain.
+class ContextBase {
+public:
+    ContextBase() = default;
+    ~ContextBase() = default;
+
+    ContextBase(const ContextBase&) = delete;
+    ContextBase(ContextBase&&) = delete;
+    ContextBase& operator=(ContextBase&&) = delete;
+    ContextBase& operator=(const ContextBase&) = delete;
+
+    // For pipeline, it is called by unref() when the last operator is unreffed.
+    // For non-pipeline, it is called by close() of the exec node directly
```

Comment:
Why non-pipeline will call this class method?

Reply:
`Aggregator` is also used by `AggregateBlockingNode` and `AggregateStreamingNode`, let me add some comment.